File: mvp_rxjava_retrofit/src/main/java/com/example/mvp_rxjava_retrofit/http/ApiException.java
package com.example.mvp_rxjava_retrofit.http;

/**
 * 自定义异常
 */
public class ApiException {
    private Throwable e;
    private int code;
    private String message;
    private int statusCode;

    public ApiException(Throwable e, int code, String message) {
        this.e = e;
        this.code = code;
        this.message = message;
    }

    public ApiException(Throwable e, int code, String message, int statusCode) {
        this.e = e;
        this.code = code;
        this.message = message;
        this.statusCode = statusCode;
    }

    public Throwable getThrowable() {
        return e;
    }

    public int getCode() {
        return code;
    }

    public String getMessage() {
        return message;
    }

    public int getStatusCode() {
        return statusCode;
    }

    /**
     * 异常状态码
     */
    public static final class Error {
        /**
         * 未知错误
         */
        public static final int UNKNOWN = 1000;
        /**
         * 解析错误
         */
        public static final int PARSE_ERROR = 1001;
        /**
         * 网络错误
         */
        public static final int NETWORD_ERROR = 1002;
        /**
         * 协议出错
         */
        public static final int HTTP_ERROR = 1003;
        /**
         * 证书出错
         */
        public static final int SSL_ERROR = 1005;
        /**
         * 无网络异常
         */
        public static final int NOT_NETWORK_ERROR = 1006;
        /**
         * 连接超时异常
         */
        public static final int CONNECT_TIMEOUT_ERROR = 1007;
        /**
         * Http状态码异常
         */
        public static final int HTTP_STATUS_ERROR = 1008;
    }
}

File: mvvm_coroutines_retrofit_flow_hilt/src/main/java/com/example/mvvm_coroutines_retrofit_flow_hilt/network/exceptions/ApiException.kt
package com.example.mvvm_coroutines_retrofit_flow_hilt.network.exceptions

/**
 * 统一管理异常类
 */
class ApiException(throwable: Throwable, val code: Int, val displayMessage: String) :
    Exception(throwable) {

    /**
     * Http状态码
     */
    object Http {
        const val UNAUTHORIZED = 401
        const val FORBIDDEN = 403
        const val NOT_FOUND = 404
        const val REQUEST_TIMEOUT = 408
        const val INTERNAL_SERVER_ERROR = 500
        const val BAD_GATEWAY = 502
        const val SERVICE_UNAVAILABLE = 503
        const val GATEWAY_TIMEOUT = 504
    }

    /**
     * 异常状态码
     */
    object Error {
        /**
         * 未知错误
         */
        const val UNKNOWN = 1000

        /**
         * 解析错误
         */
        const val PARSE_ERROR = 1001

        /**
         * 网络错误
         */
        const val NETWORD_ERROR = 1002

        /**
         * 协议出错
         */
        const val HTTP_ERROR = 1003

        /**
         * 连接超时
         */
        const val CONNECT_TIMEOUT = 1004

        /**
         * 证书出错
         */
        const val SSL_ERROR = 1005
    }
}